File: ruby/package.rb
class Package
  EOD_TIMESTAMP = Time.new(2010, 10, 10, 17)

  attr_reader :id, :street, :city, :state, :zip, :deadline, :weight_in_kilos, :delivered_at, :ready_at, :location
  attr_accessor :on_truck, :left_hub_at, :delivered_at
  
  def initialize(id, street, city, zip, deadline, weight_in_kilos, special_instructions, location)
    @id = id
    @street = street
    @city = city
    @zip = zip
    @location = location
    @deadline = convert_to_timestamp(deadline)
    @weight_in_kilos = weight_in_kilos
    @delivered_at = nil
    @ready_at = Time.new(2010, 10, 10, 8)
    @left_hub_at = nil
    @on_truck = false
    @truck_availability = [1, 2]
    @special_instructions = special_instructions
    
    modify(special_instructions)
  end

  def has_deadline?
    deadline < EOD_TIMESTAMP
  end

  def high_priority?
    has_deadline? || special_instructions != 'None' || [13, 15, 19].include?(id)
  end

  def can_be_delivered_by?(truck)
   !on_truck && truck_availability.include?(truck.id) && truck.current_time >= ready_at
  end

  def to_s
    "ID: #{id}, Location: #{street}, Deadline: #{deadline}, Ready At: #{ready_at}, Left Hub At: #{left_hub_at}, Delivered At: #{delivered_at}, Instructions: #{special_instructions}, Delivered On Time: #{delivered_at <= deadline}"
  end

  private

  attr_reader :special_instructions
  attr_writer :street, :ready_at, :zip
  attr_accessor :truck_availability

  def convert_to_timestamp(deadline)
    return Time.new(2010, 10, 10, 17) if deadline == 'EOD'

    h_m_s = deadline.split(':')
    
    hour = h_m_s[0].to_i
    min = h_m_s[1].to_i
    sec = h_m_s[2].to_i

    Time.new(2010,10,10,hour, min, sec)
  end

  def modify(instructions)
    case instructions
    when /Wrong address listed/
      self.ready_at = Time.new(2010, 10, 10, 10, 20)
      self.street = '410 S State St'
      self.zip = 84111
    when /Delayed/
      self.ready_at = Time.new(2010, 10, 10, 9, 5)
    when /Can only be on truck 2/
      self.truck_availability = [2]
    end
  end
end

File: python/package.py
from datetime import timedelta
import re

class Package(object):
    EOD_TIMESTAMP = timedelta(hours=17)
    SPECIAL_PACKAGES = [13, 15, 19]
    DELIVERED = 'Delivered At {}'
    ON_TRUCK = 'On Truck - Left Hub At: {}'
    AT_HUB = 'Waiting At Hub'

    def __init__(self, identifier, street, city, zipcode, deadline, weight_in_kilos, notes, destination):
        self.identifier = int(identifier)
        self.street = street
        self.city = city
        self.zip = zipcode
        self.deadline = self._convert_to_timestamp(deadline)
        self.weight_in_kilos = weight_in_kilos
        self.notes = notes
        self.destination = destination

        self.delivered_at = None
        self.ready_at = timedelta(hours=8)
        self.left_hub_at = None
        self.on_truck = False
        self.truck_availability = [1, 2]
        self.notes = notes

        self._modify(notes)

    def inline_report(self, time):
        report = self.report(time)

        return report[1:].replace('\n', '   ')

    def report(self, time=timedelta(hours=17)):
        return """
ID: {}
Address: {} {} UT
Zipcode: {}
Deadline: {}
Weight: {}
Delivery Status: {}\
""".format(
    self.identifier,
    self.street,
    self.city,
    self.zip,
    self._deadline(),
    self.weight_in_kilos,
    self._delivery_status(time)
)
        
    def has_deadline(self):
        return self.deadline != self.EOD_TIMESTAMP

    def _deadline(self):
        if self.deadline == self.EOD_TIMESTAMP:
            return '{} (EOD)'.format(self.deadline)

        return '{}'.format(self.deadline)

    # determines if a package is a high priority based on if it has a deadline or special instructions
    def is_high_priority(self):
        return self.has_deadline() or self.notes != 'None' or self.identifier in self.SPECIAL_PACKAGES

    # determines the delivery status based on the timestamp passed in
    def _delivery_status(self, time):
        if time > self.delivered_at:
            return self.DELIVERED.format(self.delivered_at)
        elif time > self.left_hub_at:
            return self.ON_TRUCK.format(self.left_hub_at)

        return self.AT_HUB

    def _convert_to_timestamp(self, time_string):
        if time_string == 'EOD':
            return self.EOD_TIMESTAMP

        (hour, minute, sec) = time_string.split(':')
        return timedelta(hours=int(hour), minutes=int(minute), seconds=int(sec))

    # this modifies the package state based on it's special instructions
    def _modify(self, notes):
        if re.match('Wrong address listed', notes):
            self.ready_at = timedelta(hours=10, minutes=20)
            self.street = '410 S State St'
            self.zip = 84111
        elif re.match('Delayed', notes):
            self.ready_at = timedelta(hours=9, minutes=5)
        elif re.match('Can only be on truck 2', notes):
            self.truck_availability = [2]
